### PR TITLE
feat(logging): improve error message & add link to Node errors page.

### PR DESCRIPTION
**Pull request by cloudmu**
This will add error description when logging errors. 

1. The error descriptions are "standard" descriptions supported by Node module [errno](https://www.npmjs.com/package/errno), see Node [Common System Errors](https://nodejs.org/api/errors.html#errors_class_system_error) and [POSIX errno](http://man7.org/linux/man-pages/man3/errno.3.html)

2. The error message format is also slightly modified, hopefully to be more beginner friendly. See an error example in the console of a [create-react-app](https://github.com/facebookincubator/create-react-app) test application.

hpm-error log

HPM is used by create-react-app for [Proxying API Requests](https://github.com/facebookincubator/create-react-app/blob/ef94b0561d5afb9b50b905fa5cd3f94e965c69c0/template/README.md#proxying-api-requests-in-development).   create-react-app expects more beginner friendly error messages. See [pull request](https://github.com/facebookincubator/create-react-app/pull/502).

### CODE REVIEW

**Comment by chimurai** [comment]
Hi @cloudmu,

Thanks for the PR.

I do like the improved error messaging.

You already pointed out; Not sure about the extra `errno` dependency, as it only provides a simple description for the error code.

Below I added two more variants of the error message.

Original error message:
`[HPM] PROXY ERROR: ECONNREFUSED. localhost -> http://localhost:3001/api/login`

1. PR error message:
`[HPM] Error 'connection refused' (ErrorCode=ECONNREFUSED) occurred while trying to proxy request /api/login from localhost:3000 to http://localhost:3001`

2. Variant without `errno` dependency
`[HPM] An error occurred while trying to proxy request /api/login from localhost:3000 to http://localhost:3001 (ECONNREFUSED)`

3. Variant without `errno` dependency. With link for more information on the error code.
`[HPM] An error occurred while trying to proxy request /api/login from localhost:3000 to http://localhost:3001 (ECONNREFUSED) (https://nodejs.org/api/errors.html#errors_common_system_errors)`

My preference would go to variant 3; If the goal is to make the error message less aggressive and more beginner friendly. For the following reasons:

- Uninterrupted plain English sentence explaining the error.
- Print out original error code for troubleshooting.
- Providing link for more info on the error code. (Common System Errors)
- No extra `errno` dependency

What do you think? /cc @gaearon @geowarin

**Comment by cloudmu** [comment]
@chimurai I like variant 3 too.  Very clean.

As I suggested in my original create-react-app PR, I was also thinking about a direct link to Nodejs common system error page, but wasn't sure how stable that link would stay.  Was waiting on some feedback on preferences from other people, but want to get the ball rolling.

For variant 3, I would probably still point out ECONNREFUSED to be errorcode, and put it on the same line with the link.

So variant 4:
   `[HPM] An error occurred while trying to proxy request /api/login from localhost:3000 to http://localhost:3001`
  `(ErrorCode=ECONNREFUSED. See https://nodejs.org/api/errors.html#errors_common_system_errors)`

What do you think?  /cc @gaearon @geowarin

**Comment by gaearon** [comment]
Option 3 looks good to me.

File: lib/index.js [hunk 1]
```diff
@@ -129,14 +128,10 @@ function HttpProxyMiddleware(context, opts) {
...
-        
-        if (description) {
-            logger.error('[HPM] Error \'%s\' (ErrorCode=%s) occurred while trying to proxy request %s from %s to %s', description, err.code, req.url, hostname, target);
-        } else {
-            logger.error('[HPM] Error (ErrorCode=%s) occurred while trying to proxy request %s from %s to %s', err.code, req.url, hostname, target);
-        }
+        var errReference = 'https://nodejs.org/api/errors.html#errors_common_system_errors'; // link to Node Common Systems Errors page
+
+        logger.error('[HPM] Error occurred while trying to proxy request %s from %s to %s (%s) \n(%s)', req.url, hostname, target, err.code, errReference);
     }
```
Comment (gaearon) [inline lib/index.js]: Do we really need HPM here? If this is a transitive dependency people will be confused what it means.

**Comment by cloudmu** [comment]
> Do we really need HPM here? If this is a transitive dependency people will be confused what it means.

HPM is used consistently as the source of logging for other logging levels, so it's a tough call.  Some ideas:
- option 1:  leave HPM out in the error log.
  Good for create-react-app use case, but maybe not ideal for other use cases.

- option 2:  change HPM to http-proxy-middleware.
   This makes source of logging less mysterious, and the developers can easily look it up.  (It should be done for all log levels for consistency).   Btw, http-proxy-middleware does show up as the first search result.

- option 3:  pass in logSource as argument to the module. This feels an over-kill, and complicates most uses cases. 

I like option 2. but @chimurai it's your module, and your opinion counts most :-)

**Comment by gaearon** [comment]
Option 2 sounds good to me.

**Comment by cloudmu** [comment]
@chimurai are you good with option 2 (changing HPM to http-proxy-middleware)?

**Comment by chimurai** [comment]
Option 2 looks good and it's definitely an improvement in HPM.

But I still get the feeling this won't satisfy your [original reason](https://github.com/facebookincubator/create-react-app/pull/502#issue-173534149) for creating this PR.

I think, even with this change, it is still a bit too generic and vague for the [create-react-app](https://github.com/facebookincubator/create-react-app) users.

Another option you might want to explore is to configure a custom error handler.
```javascript
var proxy = require("http-proxy-middleware");

var onErrorHandler = function (err, req, res) {
    console.log('Something went wrong.');
    console.log('And we are reporting a custom error message.');
};

var options = {target:'http://localhost:3000', onError: onErrorHandler};
```
https://github.com/chimurai/http-proxy-middleware/blob/master/recipes/proxy-events.md#onerror

With a custom `onError` handler you can really dive into the specifics, as explained by @gaearon in:
https://github.com/facebookincubator/create-react-app/pull/282#issue-168387372

This custom `onError` is not affected by the logLevels. So it's up to you to set it to 'silent' or 'error'...

File: lib/index.js [hunk 2]
```diff
@@ -128,9 +128,10 @@ function HttpProxyMiddleware(context, opts) {
...
 
     function logError(err, req, res) {
-        var hostname = (req.hostname || req.host) || (req.headers && req.headers.host); // (node0.10 || node 4/5) || (websocket)
-        var targetUri = (proxyOptions.target.host || proxyOptions.target) + req.url;
+        var hostname = (req.headers && req.headers.host) || (req.hostname || req.host);     // (websocket) || (node0.10 || node 4/5)
+        var target = proxyOptions.target.host || proxyOptions.target;
+        var errReference = 'https://nodejs.org/api/errors.html#errors_common_system_errors'; // link to Node Common Systems Errors page
 
-        logger.error('[HPM] PROXY ERROR: %s. %s -> %s', err.code, hostname, targetUri);
+        logger.error('[HPM] Error occurred while trying to proxy request %s from %s to %s (%s) \n(%s)', req.url, hostname, target, err.code, errReference);
```
Comment (chimurai) [inline lib/index.js]: Is the line-break really needed?
I can imagine when you configure `winston` in `logProvider` to log to a file; those line-breaks will add a lot of visual noise as you go through the logs...

**Comment by gaearon** [comment]
I like the `onError` idea. 👍

File: lib/index.js (hunk 2 above)
Comment (cloudmu) [inline lib/index.js]: @chimurai good point. I was only thinking console output in create-react-app.  I looked there and line-break was used to break long lines.

I agree for the logging purpose we should remove the line-break.  I can update the pull request.   I understand with your onError suggestion, this pull request won't be necessary for create-react-app use case, but the error message does look a little better overall (plus Node error page link).
Also would you like to use [http-proxy-middleware] or keep [HPM]?

**Comment by cloudmu** [comment]
I totally missed the onError in the api.  Should use that in the create-react-app use case!

File: lib/index.js (hunk 2 above)
Comment (chimurai) [inline lib/index.js]: If you can remove the line-break and keep `[HPM]` this PR is good to go!

Comment (cloudmu) [inline lib/index.js]: Done. Thanks!

**Comment by chimurai** [comment]
Thanks for the PR.

Don't think you'll need a new version asap if you are using the `onError`.
Let me know in case you do, than I'll push a new release.

**Comment by cloudmu** [comment]
@chimurai Yes, I am good. No release needed for now.

Thanks for all the discussions and tips.

**Pull request merged**